File: Task2.java
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.expressions.Window;
import org.apache.spark.sql.expressions.WindowSpec;
import org.apache.spark.sql.functions.*;
import org.apache.spark.sql.types.DataTypes;


public class FactAutoTransform {

    public static void main(String[] args)
    {
            SparkSession spark = SparkSession
                .builder()
                .appName("UserTransactions")
                .master("local[*]")
                .getOrCreate();

        Dataset<Row> transactionsDF = spark.read().format("csv").load("transactions.csv");
        Dataset<Row> usersDF = spark.read().format("csv").load("users.csv");

        Dataset<Row> userTransactionsDF = transactionsDF.join(usersDF.withColumnRenamed(user_id, userid),
                        transactionsDF.col("user_id").equalTo(usersDF.col("user_id")),
                        "inner");
        
        Dataset<Row> filteredUserTransactionsDF = UserTransactionsDF.filter(col("is_blocked").equalTo("false").and(col("is_active").equalTo(1)));


        Dataset<Row> aggregatedUserTransactionsDF = filteredUserTransactionsDF
                                .groupBy("transaction_category_id")
                                .agg(sum(col("amount")).alias("sum_amount"),
                                count(col("t.user_id")).alias("num_users")
                                );

        Dataset<Row> sortedUserTransactionsDF  = aggregatedUserTransactionsDF
                            .orderBy(col("sum_amount").desc());

        

    }



}

File: Task2.py
import csv

transactions = {}
with open('transactions.csv') as file:
    filereader = csv.DictReader(file)
    for row in filereader:
        transactions[row['transaction_id']] = row
   # print(transactions)

users = {}
with open('users.csv') as file:
    filereader = csv.DictReader(file)
    for row in filereader:
        users[row['user_id']] = row
    # print (users)

# Compute the result of the query
results = {}
for transaction_id, transaction in transactions.items():
    if transaction['is_blocked'] == 'False':
        user_id = transaction['user_id']
        if users[user_id]['is_active'] == '1':
            category_id = transaction['transaction_category_id']
            amount = transaction['amount']
            if category_id not in results:
                results[category_id] = {'sum_amount': 0, 'num_users': set()}
            results[category_id]['sum_amount'] += amount
            results[category_id]['num_users'].add(user_id)

# Print the result
for category_id, result in sorted(results.items(), key=lambda x: x[1]['sum_amount'], reverse=True):
    sum_amount = result['sum_amount']
    num_users = len(result['num_users'])
    print(f"{category_id}, {sum_amount:.2f}, {num_users}")
